### PR TITLE
Add Dimensions 'changed' event listener, expose size classes on iOS

### PR DESCRIPTION
**Pull request by jacobp100**
**Test plan (required)**

Example screenshot of iPhone 6S (which has a regular vertical size class)

screen shot 2016-09-01 at 12 06 09

Added support for dimension change events. However, the UIExplorer example will not demonstrate this, as the demo component is unmounted and remounted on screen changes. Past simple confirmation that the event works (with console.log), I'm not sure what is the best way to test this, or if it's too trivial to test.

### CODE REVIEW

**Comment by facebook-github-bot-4** [comment]
By analyzing the blame information on this pull request, we identified @ericvicenti and @javache to be potential reviewers.

**Comment by facebook-github-bot-8** [comment]
@jacobp100 updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9701/files/cc9380eca061afbdfed818f2de0b4627b2f1900a..f2a3fe30d2fccdb170fbbdc44b3816a26cae64f9)

**Comment by mkonicek** [comment]
Can the dimensions change only only on screen rotation or in other cases too?

cc @andreicoman11 who wanted to look into screen rotation on Android
cc @javache

**Comment by jacobp100** [comment]
Both platforms have some sort of split screen support.

**Comment by mkonicek** [comment]
> Both platforms have some sort of split screen support.

👍

**Comment by facebook-github-bot-8** [comment]
@jacobp100 updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9701/files/f2a3fe30d2fccdb170fbbdc44b3816a26cae64f9..0f647fa45287c0993a99d16166072d04f053a565)

File: Libraries/Utilities/Dimensions.js [hunk 1]
```diff
@@ -17,6 +17,12 @@ var RCTDeviceEventEmitter = require('RCTDeviceEventEmitter');
 
 var invariant = require('fbjs/lib/invariant');
 
+type DimensionsChangedEventName = $Enum<{
```
Comment (javache) [inline Libraries/Utilities/Dimensions.js]: `type DimensionsChangedEventName = 'backPress'`

If other events get added here, you can just add them as

`type DimensionsChangedEventName = 'backPress' | 'otherEvent'`

File: Libraries/Utilities/Dimensions.js [hunk 2]
```diff
@@ -74,13 +86,64 @@ class Dimensions {
...
+   * Dimensions.addEventListener('change', function(dimensions) {
+   *   deepEqual(dimensions.window, Dimensions.get('window')) === true
+   * });
+   * ```
+   */
+  static addEventListener(
+    eventName: DimensionsChangedEventName,
+    handler: Function
+  ): {remove: () => void} {
+    invariant(
```
Comment (javache) [inline Libraries/Utilities/Dimensions.js]: Use EventEmitter to manage listeners.

File: React/Modules/RCTUIManager.m [hunk 3]
```diff
@@ -1519,12 +1519,18 @@ static void RCTMeasureLayout(RCTShadowView *view,
   RCTAssertMainQueue();
 
   // Don't use RCTScreenSize since it the interface orientation doesn't apply to it
-  CGRect screenSize = [[UIScreen mainScreen] bounds];
+  id mainScreen = [UIScreen mainScreen];
```
Comment (javache) [inline React/Modules/RCTUIManager.m]: Use `UIScreen *` instead of `id`

**Comment by javache** [comment]
Good timing, @andreicoman11 is just adding this event on Android.

File: React/Modules/RCTUIManager.m (hunk 3 above)
Comment (littlesome) [inline React/Modules/RCTUIManager.m]: There's a bug on iPad with split screen, Dimensions.get('window') always returns screen size.

window's: width & height should be UIWindow's bounds as apple said:

> Likewise, in iOS 9 it’s no longer appropriate to employ the screen’s bounds to gauge your app’s visible area. Instead, obtain the value of your app’s window’s bounds property (which the UIWindow class inherits from the UIView class.)

[https://developer.apple.com/library/ios/documentation/WindowsViews/Conceptual/AdoptingMultitaskingOniPad/index.html#//apple_ref/doc/uid/TP40015145-CH3-SW1]()

and what about provide both window and screen like android do?

**Comment by mkonicek** [comment]
> Good timing, @andreicoman11 is just adding this event on Android.

Nice!

**Comment by jacobp100** [comment]
Friendly ping to @andreicoman11 or @javache

File: Libraries/Utilities/Dimensions.js (hunk 1 above)
Comment (hramos) [inline Libraries/Utilities/Dimensions.js]: @jacobp100 see feedback above.

File: Libraries/Utilities/Dimensions.js (hunk 2 above)
Comment (hramos) [inline Libraries/Utilities/Dimensions.js]: @jacobp100 see feedback above.

File: React/Modules/RCTUIManager.m (hunk 3 above)
Comment (hramos) [inline React/Modules/RCTUIManager.m]: @jacobp100 see @javache feedback above.

**Comment by jacobp100** [comment]
Sorry, I somehow missed all the comments. Added the suggestions, but I wasn’t sure what the standard way to do event emitters in RN is.

File: Libraries/Utilities/Dimensions.js [hunk 4]
```diff
@@ -11,17 +11,18 @@
...
 'use strict';
 
+var NativeEventEmitter = require('EventEmitter');
 var Platform = require('Platform');
 var UIManager = require('UIManager');
 var RCTDeviceEventEmitter = require('RCTDeviceEventEmitter');
 
 var invariant = require('fbjs/lib/invariant');
 
 type DimensionsChangedEventName = $Enum<{
```
Comment (ericvicenti) [inline Libraries/Utilities/Dimensions.js]: Could you do this instead?

```javascript
type DimensionsEventName = 'change';
// eventually maybe this:
type DimensionsEventName = 'change' | 'otherEvent';
```

**Comment by lacker** [comment]
Hey, are you still working on this pull request? In terms of feedback, it seems pretty risky to add this functionality without any tests that cover it. Could you arrange so that this dimension behavior is tested? We also would need to document it.

**Comment by jacobp100** [comment]
Sorry, I didn’t get a notification for the change requests. I’m not too sure how we do tests, but for the size classes, there is an example in UIExplorer that will at least form a screenshot-based test.

**Comment by jacobp100** [comment]
@hramos Did you see my previous comment?

**Comment by jacobp100** [comment]
So it looks like the event emitter part was added to master at some point, which makes this PR just about the size classes now! I've just updated to master and tested everything.

Is there any more I need to do? I believe the only issues were in `Dimensions.js`, which is now the master version.

**Comment by jacobp100** [comment]
Friendly ping to @lacker or @hramos

**Comment by shergin** [comment]
@jacobp100 Thank you so much for your effort!

I believe we should not introduce `size class` concept (and merge this) because:
 * This is pretty specific iOS functionality and I can not imagine how it can be used in RN;
 * `size class` is actually `view`/`view controller` property, not just `screen`/`device` property, so this implementation is not quite correct/useful.

Please, feel free to reopen this to continue the discussion if you are disagree.

**Comment by jacobp100** [comment]
Makes sense. Definitely valid points!

**Pull request closed without merging**